docs: CAPI docs review (#957)
canonical/k8s-snap: docs/src/capi/howto/custom-ck8s.md

# Install custom {{product}} on machines

By default, the `version` field in the machine specifications will determine 
which {{product}} version is downloaded from the `stable` risk level. While 
you can install different versions of the `stable` risk level by changing the 
`version` field, extra steps should be taken if you're willing to install 

> a specific risk level, revision, or install the snap from local path.

Reviewer: Possibly break this sentence in two or make it a bit shorter. Maybe we can point people to the channels page instead?

Author: Changed it to:

By default, the `version` field in the machine specifications will determine 
which {{product}} **version** is downloaded from the `stable` risk level.
This guide walks you through the process of installing {{product}} 
with a specific **risk level**, **revision**, or from a **local path**.

Hopefully it's a bit better now.

This guide walks you through the process of installing custom {{product}} 
on workload cluster machines.

## Prerequisites

To follow this guide, you will need:

- A Kubernetes management cluster with Cluster API and providers installed 
and configured.
- A generated cluster spec manifest

Please refer to the [getting-started guide][getting-started] for further
details on the required setup.

This guide will call the generated cluster spec manifest `cluster.yaml`.

> ## Using config spec

Reviewer: ```suggestion
## Using the configuration specification
```
Let's avoid abbreviations.
Maybe I am lacking context but what spec is this?

{{product}} can be installed on machines using a specific `channel`, 
`revision` or `localPath` by specifying the respective field in the spec 
of the machine.

> ```yaml

Reviewer 2: I would still introduce what this file is

Author: I think GitHub diff annotations are a bit misleading here. The `Overwrite the existing `install.sh` script` is still there, but it's now the second option rather than the first one. I'll add some info around what this file is in that section.

spec:
  ...
  spec:

>     channel: 1.32-classic/candidate

Reviewer: ```suggestion
    channel: 1.xx-classic/candidate
```
Should we stay generic here?

# Install custom {{product}} on machines

By default, the `version` field in the machine specifications will determine 

> which {{product}} is downloaded from the `stable` risk level. While you can 

Reviewer 3: I don't think this is clear: "Which Canonical Kubernetes is downloaded from the stable risk level".

Maybe you could add the word "version"? "Which {{product}} *version* is downloaded... "

Suggestion: "you can adjust the version field to install other versions within the stable risk level, additional steps are required if you wish to install a specific risk level."

Author: I changed this a bit to also include revision and local path.

`version` field, extra steps should be taken if you're willing to install 
a specific risk level.
This guide walks you through the process of installing custom {{product}} 
on workload cluster machines.

## Prerequisites

To follow this guide, you will need:

- A Kubernetes management cluster with Cluster API and providers installed 
and configured.
- A generated cluster spec manifest

Please refer to the [getting-started guide][getting-started] for further
details on the required setup.

This guide will call the generated cluster spec manifest `cluster.yaml`.

## Using Config Spec

{{product}} can be installed on machines using a specific `channel`, 
`revision` or `localPath` by specifying the respective field in the spec 
of the machine.

```yaml
spec:
  ...
  spec:
    channel: 1.32-classic/candidate
    # Or
    revision: 1234
    # Or
    localPath: /path/to/snap/on/machine
```

Note that for the `localPath` to work the snap must be available on the 
machine at the specified path on boot.

## Overwrite the existing `install.sh` script

The installation of the {{product}} snap is done via running the `install.sh` 

> script in the cloud-init.

Reviewer 3: Technically, cloud-init reads it's userdata string and runs the install.sh command.

Author: Does this change cut it?
```suggestion
The installation of the {{product}} snap is done via running the `install.sh` 
script by the cloud-init.
```

Reviewer 3: Sure, thanks!

While this file is automatically placed in every workload cluster machine 

> which hard-coded content by {{product}} providers, you can overwrite this 

Reviewer 3: ```suggestion
with hard-coded content by {{product}} providers, you can overwrite this 
```

> file to make sure your desired content is available in the script.

Reviewer 3: ```suggestion
file suit your needs.
```

> As an example, let's overwrite the `install.sh` for our control plane nodes. 

Reviewer 3: This is an OK example, but I hope people won't get the wrong message here. They don't need to override install.sh to specify the track or risk level, there are fields in CK8sConfig CRD for these. See [channel](https://github.com/canonical/cluster-api-k8s/blob/cdc21cb63ea7f4287f174f878ae2d11045f010ae/bootstrap/api/v1beta2/ck8sconfig_types.go#L86)

Author: That's right, I included this as the first option as well. Hopefully the get that they don't need this. Let me also add a line about it.

Now the new control plane nodes that are created using this manifest will have
the `1.31-classic/candidate` {{product}} snap installed on them!

```{note}

> If you're only interested in installing a specific channel, revision, or 
> form the local path, it's recommended to 
> [use the config spec](#using-config-spec).

Reviewer: ```suggestion
[Use the configuration specification](#using-config-spec), if you're only interested in installing a specific channel, revision, or 
form the local path.
```
Generally things sound better in the conditional if you switch things around like this:

> ## Use `preRunCommands`

Reviewer 3: There are also postRunCommands, but these aren't documented here?

Author: I think it's best to remove this option overall. It's extremely hacky and probably not a good fit for a how-to.

> As mentioned above, the `install.sh` script is responsible for installing 

Reviewer 4: This feels a bit too much of an explanation than a how-to. 
It explains what each script does rather than in which case you would overwrite it. 
Please either provide an example or remove this section.

Author: This section is removed.

{{product}} snap on machines. `preRunCommands` are executed before 
`install.sh`. You can also add an install command to the `preRunCommands` 
in order to install your desired {{product}} version.

```{note}
Installing the {{product}} snap via the `preRunCommands`, does not prevent 
the `install.sh` script from running. Instead, the installation process in 
the `install.sh` will fail with a message indicating that `k8s` is already 
installed.
This is not considered a standard way and overwriting the `install.sh` 

> script is recommended.

Reviewer 3: This might be the right place to mention that they don't need to overwrite install.sh at all, in fact they should set [channel](https://github.com/canonical/cluster-api-k8s/blob/cdc21cb63ea7f4287f174f878ae2d11045f010ae/bootstrap/api/v1beta2/ck8sconfig_types.go#L86) instead. I don't know why we are recommending overwriting install.sh at all.

Author: Maybe we can remove this part all together?
